### PR TITLE
Various refactors related to the notification system

### PR DESCRIPTION
**Pull request by jon-betts**
These are intended to clear the way for:

 * https://github.com/hypothesis/h/pull/7741

This will hopefully make the changes there smaller and more obvious

## Review notes

This PR is split into many granular commits to try and make it easy to review and verify correctness.

### CODE REVIEW

File: h/services/user_signup.py [hunk 1]
```diff
@@ -113,18 +110,20 @@ def _require_activation(self, user):
...
-        mail_params = self.signup_email(
-            user_id=user.id, email=user.email, activation_code=user.activation.code
+        mail_params = signup.generate(
+            request=self.request,
+            user_id=user.id,
+            email=user.email,
+            activation_code=user.activation.code,
         )
-        self.mailer.send.delay(*mail_params)
+        tasks_mailer.send.delay(*mail_params)
```
Comment (jon-betts) [inline h/services/user_signup.py:119]: I think all of this injection happening here was intended to make testing easier, but I think it kind of did the reverse by:

 * Bloating the interface
 * Causing us to have to pass in weird things like partials which are hard to test and mock

This is the only way this is ever configured to run, and the extra degrees for freedom are just confusing and indirect.

File: tests/common/factories/__init__.py [hunk 2]
```diff
@@ -1,47 +1,20 @@
...
+from tests.common.factories.feature import Feature
+from tests.common.factories.flag import Flag
+from tests.common.factories.group import Group, OpenGroup, RestrictedGroup
+from tests.common.factories.group_scope import GroupScope
+from tests.common.factories.job import Job, SyncAnnotationJob
+from tests.common.factories.organization import Organization
+from tests.common.factories.setting import Setting
+from tests.common.factories.token import DeveloperToken, OAuth2Token
+from tests.common.factories.user import User
+from tests.common.factories.user_identity import UserIdentity
```
Comment (jon-betts) [inline tests/common/factories/__init__.py:20]: This more closely represents our new style I think

File: tests/common/fixtures/services.py [hunk 3]
```diff
@@ -44,17 +49,13 @@
...
     "user_service",
     "user_signup_service",
     "user_unique_service",
     "user_update_service",
 )
 
-from h.services.user import UserService
-from h.services.user_signup import UserSignupService
-from h.services.user_unique import UserUniqueService
-from h.services.user_update import UserUpdateService
```
Comment (jon-betts) [inline tests/common/fixtures/services.py:56]: Howdy fella. You lost?

File: tests/h/services/user_signup_test.py [hunk 4]
```diff
@@ -165,75 +166,34 @@ def test_signup_raises_conflict_error_when_account_already_exists(
...
-    def password_setter(user, password):  # pylint:disable=unused-argument
-        user.password = "fakehash"
-
-    service.update_password.side_effect = password_setter
-
-    pyramid_config.register_service(service, name="user_password")
-    return service
+    @pytest.fixture
+    def UserSignupService(self, patch):
+        return patch("h.services.user_signup.UserSignupService")
```
Comment (jon-betts) [inline tests/h/services/user_signup_test.py:199]: This diff is a giant mess, but the original test for the factory was in the style where we had a different test for each argument we passed to the service. 

Allow me a quite whine. This is too much and I'd say is actively worse than a single test for a few reasons:

 * General bloat - the tests are out of proportion to the code tested
 * With a single test you have to get the call right, with many you can forget a bit
 * It's hard to look at the tests and see what they are testing

Other oddities here include mocking things we shouldn't, having active mocks which try to "do" things rather than relying on `return_value` and not mocking the actual service.

If we want to assert behavior about the init of the service I think that should go in the service tests?

File: tests/h/views/accounts_test.py [hunk 5]
```diff
@@ -688,97 +688,71 @@ def routes(self, pyramid_config):
...
-        assert result["user_has_email_address"] is None
-
-    def test_it_renders_form_if_user_has_email_address(
-        self, factories, form_validating_to, pyramid_request
-    ):
-        controller = views.NotificationsController(pyramid_request)
-        controller.form = form_validating_to({})
-        pyramid_request.user = factories.User(username="janedoe")
-
-        result = controller.get()
```
Comment (jon-betts) [inline tests/h/views/accounts_test.py:730]: Some of the testing style where we separate assertions about what the method calls and what it returns etc. Sometimes you see a different test for each thing the methods calls.

This causes the setup to be duplicated over and over and makes the flow hard to follow.

File: tests/h/views/accounts_test.py [hunk 6]
```diff
@@ -688,97 +688,71 @@ def routes(self, pyramid_config):
...
-        pyramid_request.user = factories.User(username="cara")
-        pyramid_request.user.email = None
-        pyramid_config.testing_securitypolicy("jerry")
-        controller = views.NotificationsController(pyramid_request)
-        controller.form = invalid_form()
+    def test_get_if_user_has_no_email_address(self, controller, authenticated_user):
+        authenticated_user.email = None
 
-        result = controller.post()
+        result = controller.get()
```
Comment (jon-betts) [inline tests/h/views/accounts_test.py:709]: I think this is due to re-ordering the tests to be:

 * test_get - happy
 * test_get - sad
 * test_post - happy
 * test_post - sad

File: tests/h/views/accounts_test.py [hunk 7]
```diff
@@ -688,97 +688,71 @@ def routes(self, pyramid_config):
...
+
+        return authenticated_user
+
+    @pytest.fixture(autouse=True)
     def routes(self, pyramid_config):
         pyramid_config.add_route("account_notifications", "/p/notifications")
 
+    @pytest.fixture(autouse=True)
+    def subscriptions_model(self, patch):
+        return patch("h.models.Subscriptions")
```
Comment (jon-betts) [inline tests/h/views/accounts_test.py:754]: No other tests use this fixture. In a test file 900 lines long this is really unhelpful to split up. It means surfing hundreds of lines to try and see the relevant fixtures.

It also makes it a pain if you come to move this into separate files at some point. If the tests are more self contained, you can lift and shift them with much less drama.

**Review by marcospri — APPROVED** [review]

**Comment by marcospri** [comment]
Looks good, tested this directly on https://github.com/hypothesis/h/pull/7741
